### PR TITLE
Corrected dependencies with cpp-dependencies

### PR DESCRIPTION
**Pull request by bilke**
See https://github.com/tomtom-international/cpp-dependencies

## Example output:

```bash
$ cpp-dependencies --stats .
21 components with 42 public dependencies, 59 private dependencies
Detected 7 nodes in cycles
```

Reveals some circular dependencies, e.g.

```bash
$ cpp-dependencies --shortest MathLib MeshLib
MathLib -> MeshLib
  ./MathLib/LinAlg/MatrixVectorTraits.cpp includes ./MeshLib/NodePartitionedMesh.h

$ cpp-dependencies --shortest MeshLib MathLib
MeshLib -> MathLib
  ./MeshLib/MeshGenerators/MeshLayerMapper.cpp includes ./MathLib/MathTools.h
  ./MeshLib/Node.h includes ./MathLib/Vector3.h
  ./MeshLib/Node.h includes ./MathLib/Point3dWithID.h
  ./MeshLib/MeshGenerators/LayeredVolume.cpp includes ./MathLib/Vector3.h
  ./MeshLib/MeshQuality/AngleSkewMetric.cpp includes ./MathLib/MathTools.h
  ./MeshLib/MeshEditing/MeshRevision.cpp includes ./MathLib/GeometricBasics.h
  ./MeshLib/MeshSurfaceExtraction.h includes ./MathLib/Vector3.h
  ./MeshLib/Elements/TriRule3.cpp includes ./MathLib/GeometricBasics.h
  ./MeshLib/Elements/TetRule4.cpp includes ./MathLib/GeometricBasics.h
  ./MeshLib/Elements/PyramidRule5.cpp includes ./MathLib/GeometricBasics.h
  ./MeshLib/Elements/QuadRule4.cpp includes ./MathLib/GeometricBasics.h
  ./MeshLib/Elements/FaceRule.h includes ./MathLib/Vector3.h
  ./MeshLib/Elements/HexRule8.cpp includes ./MathLib/GeometricBasics.h
  ./MeshLib/Elements/Element.cpp includes ./MathLib/GeometricBasics.h
  ./MeshLib/Elements/Element.cpp includes ./MathLib/MathTools.h
  ./MeshLib/Elements/PrismRule6.cpp includes ./MathLib/GeometricBasics.h
  ./MeshLib/Elements/CellRule.cpp includes ./MathLib/Vector3.h
  ./MeshLib/Elements/Element.h includes ./MathLib/Point3d.h
  ./MeshLib/MeshSearch/MeshElementGrid.h includes ./MathLib/Point3d.h
  ./MeshLib/MeshGenerators/MeshGenerator.h includes ./MathLib/Point3d.h
  ./MeshLib/Elements/PointRule1.cpp includes ./MathLib/Point3d.h
  ./MeshLib/ElementCoordinatesMappingLocal.h includes ./MathLib/Point3d.h
  ./MeshLib/Elements/TemplateElement.h includes ./MathLib/Point3d.h
  ./MeshLib/MeshEditing/projectMeshOntoPlane.h includes ./MathLib/Vector3.h
  ./MeshLib/MeshEditing/projectMeshOntoPlane.h includes ./MathLib/Point3d.h
  ./MeshLib/MeshEditing/projectMeshOntoPlane.h includes ./MathLib/MathTools.h
  ./MeshLib/Elements/FaceRule.cpp includes ./MathLib/MathTools.h
  ./MeshLib/CoordinateSystem.h includes ./MathLib/Vector3.h
  ./MeshLib/ElementCoordinatesMappingLocal.cpp includes ./MathLib/Vector3.h
  ./MeshLib/ElementCoordinatesMappingLocal.cpp includes ./MathLib/MathTools.h
  ./MeshLib/ElementCoordinatesMappingLocal.cpp includes ./MathLib/Point3d.h
  ./MeshLib/MeshQuality/EdgeRatioMetric.h includes ./MathLib/Point3d.h
  ./MeshLib/MeshQuality/EdgeRatioMetric.cpp includes ./MathLib/MathTools.h
```

--> The single MathLib to MeshLib dependency should be removed. :white_check_mark: 

New output (down to 4 nodes in cycles):

```bash
$ cpp-dependencies --stats .

21 components with 42 public dependencies, 58 private dependencies
Detected 4 nodes in cycles
```

## Graph generation:

```bash
$ cpp-dependencies --graph dependencies.dot; dot -Tpng dependencies.dot
>dependencies.png
```

### Libs:
dependencies

### All:
dependencies

## Future TODOs

- [ ] Fix circular dependencies
- [ ] Check circular dependencies in Jenkins, fail the build if found

### CODE REVIEW

File: Applications/DataExplorer/DataView/CMakeLists.txt [hunk 1]
```diff
@@ -134,7 +134,7 @@ add_library(QtDataView
 )
 
 target_link_libraries(QtDataView
-    PUBLIC GeoLib MeshGeoToolsLib MeshLib QtBase VtkVis
+    PUBLIC GeoLib MeshGeoToolsLib MeshLib QtBase # VtkVis
```
Comment (endJunction) [inline Applications/DataExplorer/DataView/CMakeLists.txt]: Drop them, don't comment, but probably you would do this anyway, just reminding...

File: .gitignore [hunk 2]
```diff
@@ -17,3 +17,5 @@ conaninfo.txt
 CMakeLists.txt.user*
 
 *.pyc
+
+dependencies.*
```
Comment (endJunction) [inline .gitignore]: Is it possible to create these files in build dir?

File: config-cpp-dependencies.txt [hunk 3]
```diff
@@ -0,0 +1,45 @@
+# Configuration file for cpp-dependencies
+# This is the example file that has all the default values listed.
+
+# Version that was current when this configuration file was written
+# Used to prevent a newer version with possibly incompatible generation from overwriting
+# cmakefiles automatically.
+versionUsed: 2
+
+# Company name to use in generated CMakeLists' copyright statement.
+companyName: YourCompany
```
Comment (endJunction) [inline config-cpp-dependencies.txt]: OpenGeoSys Community (www.opengeosys.org)

**Comment by endJunction** [comment]
Cool stuff! Would it be possible to check for circular dependencies automatically (and fail the build eventually)?

**Comment by endJunction** [comment]
Fixing MaterialLib<->ProcessLib is not simple. `ProcessLib::Parameter`s are needed for look-up of values in many material models.  We could move `Parameter`s in another library, or create an independent interface for look-up (_e.g._ passing `std::function`)... This is to be discussed.

File: MeshLib/CMakeLists.txt [hunk 4]
```diff
@@ -27,7 +27,7 @@ if(OGS_USE_PCH)
 endif()
 
 target_link_libraries(MeshLib
-    PUBLIC BaseLib GeoLib MathLib
+    PUBLIC BaseLib GeoLib MathLib logog
```
Comment (TomFischer) [inline MeshLib/CMakeLists.txt]: Just for understanding, why is the logog not private?

**Pull request merged**